app/views/papers/_sort_by.html.erb

```diff
@@ -0,0 +1,4 @@
+<div class="btn-group btn-group-xs nav-button-group sort" role="group" aria-label="...">
+  <%= link_to 'Votes', page == 'list' ? list_path(sort: 'votes') : papers_path(sort: 'votes'), class: "btn btn-default" + (params[:sort] == "pub_date" ? "" : " active"), role: "button" %>
+  <%= link_to 'Publish Date', page == 'list' ? list_path(sort: 'pub_date') : papers_path(sort: 'pub_date'), class: "btn btn-default" + (params[:sort] == "pub_date" ? " active" : ""), role: "button"  %>
```

Comment:
Can you redirect to self (the originating route) but with the additional params? That way you don't need to switch based on whether it's a list or a page route.